File: 3720-lexicographically-smallest-permutation-greater-than-target/3720-lexicographically-smallest-permutation-greater-than-target.py
class Solution:
    def lexGreaterPermutation(self, s: str, target: str) -> str:
        cnt = [0] * 26

        for c in s:
            cnt[ord(c) - ord('a')] += 1

        n = len(target)

        for i in range(n):
            x = ord(target[i]) - ord('a')

            if cnt[x] == 0:
                break

            cnt[x] -= 1
        else:
            i = n

        for j in range(i, -1, -1):
            if j < i:
                cnt[ord(target[j]) - ord('a')] += 1

            x = ord(target[j]) - ord('a') if j < n else -1

            for k in range(x + 1, 26):
                if cnt[k] > 0:
                    ans = target[:j] + chr(k + ord('a'))
                    cnt[k] -= 1

                    for c in range(26):
                        ans += chr(c + ord('a')) * cnt[c]

                    return ans

        return ""

File: 3720-lexicographically-smallest-permutation-greater-than-target/3720-lexicographically-smallest-permutation-greater-than-target.cpp
class Solution {
public:
    string lexGreaterPermutation(string s, string target) {
        vector<int> sc(26), pc(26);
        
        for (char c : s)
            sc[c - 'a']++;
        
        int n = target.size();
        
        for (int i = 0; i < n; i++)
            pc[target[i] - 'a']++;
        
        for (int i = n - 1; i >= 0; i--) {
            pc[target[i] - 'a']--;
            
            bool possible = true;
            for (int j = 0; j < 26; j++) {
                if (pc[j] > sc[j]) {
                    possible = false;
                    break;
                }
            }
            
            if (!possible)
                continue;
            
            int x = target[i] - 'a';
            
            for (int c = x + 1; c < 26; c++) {
                if (sc[c] > pc[c]) {
                    string ans = target.substr(0, i);
                    ans += char('a' + c);
                    
                    vector<int> rem = sc;
                    
                    for (int j = 0; j < i; j++)
                        rem[target[j] - 'a']--;
                    
                    rem[c]--;
                    
                    for (int j = 0; j < 26; j++)
                        ans += string(rem[j], char('a' + j));
                    
                    return ans;
                }
            }
        }
        
        return "";
    }
};